Mathlib/NumberTheory/LSeries/Dirichlet.lean

```diff
@@ -0,0 +1,416 @@
+
+We also prove some general results on L-series associated to Dirichlet characters
+(i.e., Dirichlet L-series). For example, we show that the abscissa of absolute convergence
+equals `1` (see `DirichletCharacter.absicssaOfAbsConv`) and that the L-series does not
+vanish on the open half-plane `re s > 1` (see `DirichletCharacter.LSeries_ne_zero_of_one_lt_re`).
+
+We deduce results on the Riemann zeta function (which is `L 1` or `L ↗ζ` on `re s > 1`)
+as special cases.
+
+## Tags
+
+Dirichlet L-series, Möbius function, von Mangoldt function, Riemann zeta function
+-/
+
+open scoped LSeries.notation
+
+/-- `δ` is the function underlying the arithmetic function `1`. -/
+lemma ArithmeticFunction.one_eq_delta : ↗(1 : ArithmeticFunction ℂ) = δ := by
+  ext
+  simp only [one_apply, LSeries.delta]
+
+section Moebius
+
+/-!
+### The L-series of the Möbius function
+
+We show that `L μ s` converges absolutely if and only if `re s > 1`.
+-/
+
+namespace ArithmeticFunction
+
+open LSeries Nat Complex
+
+lemma not_LSeriesSummable_moebius_at_one : ¬ LSeriesSummable ↗μ 1 := by
+  intro h
+  refine not_summable_one_div_on_primes <| summable_ofReal.mp <| Summable.of_neg ?_
+  simp only [← Pi.neg_def, Set.indicator_comp_of_zero ofReal_zero, ofReal_inv, ofReal_nat_cast]
+  convert h.indicator {n | n.Prime} using 1
+  ext n
+  by_cases hn : n ∈ {p | p.Prime}
+  · simp only [Pi.neg_apply, Set.indicator_of_mem hn, ofReal_div, ofReal_one, ofReal_nat_cast,
+      neg_div', term_of_ne_zero hn.ne_zero, moebius_apply_prime hn, Int.reduceNeg, Int.cast_neg,
+      Int.cast_one, cpow_one]
+  · simp only [one_div, Pi.neg_apply, Set.indicator_of_not_mem hn, ofReal_zero, neg_zero]
+
+/-- The L-series of the Möbius function converges absolutely at `s` if and only if `re s > 1`. -/
+lemma LSeriesSummable_moebius_iff {s : ℂ} : LSeriesSummable ↗μ s ↔ 1 < s.re := by
+  refine ⟨fun H ↦ ?_, LSeriesSummable_of_bounded_of_one_lt_re (m := 1) fun n ↦ ?_⟩
+  · by_contra! h
+    have h' : s.re ≤ (1 : ℂ).re := by simp only [one_re, h]
+    exact not_LSeriesSummable_moebius_at_one <| LSeriesSummable.of_re_le_re h' H
+  · rw [abs_intCast, ← Int.cast_abs]
+    norm_cast
+    rcases eq_or_ne (μ n) 0 with h | h
+    · simp [h]
+    · rcases moebius_ne_zero_iff_eq_or.mp h with h | h <;> simp [h]
+
+/-- The abscissa of absolute convergence of the L-series of the Möbius function is `1`. -/
+lemma abscissaOfAbsConv_moebius : abscissaOfAbsConv ↗μ = 1 := by
+  simpa only [abscissaOfAbsConv, LSeriesSummable_moebius_iff, ofReal_re, Set.Ioi_def,
+    EReal.image_coe_Ioi, EReal.coe_one] using csInf_Ioo <| EReal.coe_lt_top _
+
+end ArithmeticFunction
+
+end Moebius
+
+/-!
+### L-series of Dirichlet characters
+-/
+
+open Nat
+
+open scoped ArithmeticFunction.zeta in
+lemma ArithmeticFunction.const_one_eq_zeta {R : Type*} [Semiring R] :
+    ∀ n ≠ 0, (1 : ℕ → R) n = (ζ ·) n := by
+  intro _ hn
+  simp only [Pi.one_apply, zeta_apply, hn, ↓reduceIte, cast_one]
```

Comment:
I'd go for
```suggestion
lemma ArithmeticFunction.const_one_eq_zeta {R : Type*} [Semiring R] (n : ℕ) (hn : n ≠ 0) :
    (1 : ℕ → R) n = (ζ ·) n := by
  simp only [Pi.one_apply, zeta_apply, hn, ↓reduceIte, cast_one]
```
but I'd also make `n` implicit - this would probably require doing the same to the arguments to `convolution_congr`

Reply:
OK. For consistency, I've adapted the arguments of all congruence lemmas involving L-series and related stuff. (But it would be even better if `∀ n ≠ 0, ...` would read as `∀ {n} (_ : n ≠ 0), ...`.)